File: TEST_PRM_SE184160 - Copy/app/src/main/java/com/example/khanghvse184160/api/CreateOrder.java
package com.example.khanghvse184160.api;

import com.example.khanghvse184160.constant.AppInfo;
import com.example.khanghvse184160.helper.Helpers;

import org.json.JSONObject;

import java.util.Date;

import okhttp3.FormBody;
import okhttp3.RequestBody;

public class CreateOrder {
    private class CreateOrderData {
        String app_id;
        String app_user;
        String app_time;
        String amount;
        String app_trans_id;
        String embed_data;
        String item;
        String bank_code;
        String description;
        String mac;
        String callback_url;

        private CreateOrderData(String amount) throws Exception {
            long appTime = new Date().getTime();
            app_id = String.valueOf(AppInfo.APP_ID);
            app_user = "Android_Demo";
            app_time = String.valueOf(appTime);
            this.amount = amount;
            app_trans_id = Helpers.getAppTransId();
            embed_data = "{}";
            item = "[]";
            bank_code = "zalopayapp";
            description = "Payment for order #" + Helpers.getAppTransId();
            callback_url = AppInfo.CALLBACK_URL;
            
            // MAC for v2: app_id|app_trans_id|app_user|amount|app_time|embed_data|item
            String inputHMac = String.format("%s|%s|%s|%s|%s|%s|%s",
                    this.app_id,
                    this.app_trans_id,
                    this.app_user,
                    this.amount,
                    this.app_time,
                    this.embed_data,
                    this.item);

            mac = Helpers.getMac(AppInfo.MAC_KEY, inputHMac);
        }
    }

     public JSONObject createOrder(String amount) throws Exception {
        CreateOrderData input = new CreateOrderData(amount);

        RequestBody formBody = new FormBody.Builder()
                .add("app_id", input.app_id)
                .add("app_user", input.app_user)
                .add("app_time", input.app_time)
                .add("amount", input.amount)
                .add("app_trans_id", input.app_trans_id)
                .add("embed_data", input.embed_data)
                .add("item", input.item)
                .add("bank_code", input.bank_code)
                .add("description", input.description)
                .add("callback_url", input.callback_url)
                .add("mac", input.mac)
                .build();

        JSONObject data = HttpProvider.sendPost(AppInfo.URL_CREATE_ORDER, formBody);
        return data;
    }
}

File: app/src/main/java/com/example/evsecondhand/data/zalopay/CreateOrder.kt
package com.example.evsecondhand.data.zalopay

import okhttp3.FormBody
import org.json.JSONObject
import java.util.Date

class CreateOrder {
    
    private data class CreateOrderData(
        val appId: String,
        val appUser: String,
        val appTime: String,
        val amount: String,
        val appTransId: String,
        val embedData: String,
        val item: String,
        val bankCode: String,
        val description: String,
        val mac: String,
        val callbackUrl: String
    ) {
        companion object {
            fun create(amount: String): CreateOrderData {
                val appTime = Date().time
                val appId = ZaloPayConfig.APP_ID.toString()
                val appUser = ZaloPayConfig.APP_USER
                val appTransId = ZaloPayHelper.getAppTransId()
                val embedData = "{}"
                val item = "[]"
                val bankCode = ZaloPayConfig.BANK_CODE
                val description = "EV Market - Nạp tiền vào ví #$appTransId"
                val callbackUrl = ZaloPayConfig.CALLBACK_URL
                
                // MAC for v2: app_id|app_trans_id|app_user|amount|app_time|embed_data|item
                val inputHMac = String.format(
                    "%s|%s|%s|%s|%s|%s|%s",
                    appId,
                    appTransId,
                    appUser,
                    amount,
                    appTime.toString(),
                    embedData,
                    item
                )
                
                val mac = ZaloPayHelper.getMac(ZaloPayConfig.MAC_KEY, inputHMac)
                
                return CreateOrderData(
                    appId = appId,
                    appUser = appUser,
                    appTime = appTime.toString(),
                    amount = amount,
                    appTransId = appTransId,
                    embedData = embedData,
                    item = item,
                    bankCode = bankCode,
                    description = description,
                    mac = mac,
                    callbackUrl = callbackUrl
                )
            }
        }
    }
    
    fun createOrder(amount: String): JSONObject? {
        val orderData = CreateOrderData.create(amount)
        
        val formBody = FormBody.Builder()
            .add("app_id", orderData.appId)
            .add("app_user", orderData.appUser)
            .add("app_time", orderData.appTime)
            .add("amount", orderData.amount)
            .add("app_trans_id", orderData.appTransId)
            .add("embed_data", orderData.embedData)
            .add("item", orderData.item)
            .add("bank_code", orderData.bankCode)
            .add("description", orderData.description)
            .add("callback_url", orderData.callbackUrl)
            .add("mac", orderData.mac)
            .build()
        
        return HttpProvider.sendPost(ZaloPayConfig.CREATE_ORDER_URL, formBody)
    }
}
